xapian-letor/ranker/listmle_ranker.cc

```diff
@@ -0,0 +1,211 @@
+	vector<double> feature_sets = sorted_feature_vectors[i].get_fvals();
+	for (size_t j = 0; j < feature_sets.size() - 1; ++j) {
+	    gradient[j] += feature_sets[j] * exponents[i] / expsum;
+	}
+    }
+
+    vector<double> first_place_in_ground_truth_feature_sets =
+	    sorted_feature_vectors[0].get_fvals();
+
+    for (size_t i = 0; i < first_place_in_ground_truth_feature_sets.size() - 1;
+	 ++i) {
+	gradient[i] -= first_place_in_ground_truth_feature_sets[i];
+    }
+
+    return gradient;
+}
+
+static void
+update_parameters(vector<double>& new_parameters,
+		  const vector<double>& gradient,
+		  const double& learning_rate)
+{
+    size_t num = new_parameters.size();
+    for (size_t i = 0; i < num; ++i) {
+	new_parameters[i] -= gradient[i] * learning_rate;
+    }
+}
+
+/* Updates new_parameters after calculating the gradient.
+ * new_parameters(w) is updated as: w = w - gradient * learningRate
+ */
+static void
+batch_learning(vector<FeatureVector> feature_vectors,
+	       vector<double> & new_parameters,
+	       const double& learning_rate)
+{
+    sort(feature_vectors.begin(), feature_vectors.end(), label_comparer);
+    vector<double> gradient = calculate_gradient(feature_vectors,
+						new_parameters);
+    update_parameters(new_parameters, gradient, learning_rate);
+}
+
+void
+ListMLERanker::train(const vector<FeatureVector>& training_data)
+{
+    LOGCALL_VOID(API, "ListMLERanker::train", training_data);
+    if (training_data.empty()) {
+	throw Xapian::InvalidArgumentError("Training data is empty."
+					   "Check the training file");
+    }
+    int feature_cnt = training_data[0].get_fcount();
+
+    //initialize the parameters for neural network
+    vector<double> new_parameters;
+    for (int feature_num = 0; feature_num < feature_cnt; ++feature_num) {
+	new_parameters.push_back(0.0);
+    }
+
+    for (int iter_num = 1; iter_num < iterations; ++iter_num) {
+	batch_learning(training_data, new_parameters, learning_rate);
+    }
+
+    swap(parameters, new_parameters);
+}
+
+void
+ListMLERanker::save_model_to_metadata(const string& model_key)
+{
+    LOGCALL_VOID(API, "ListMLERanker::save_model_to_metadata", model_key);
+    WritableDatabase letor_db(get_database_path());
+    string key = model_key;
+    if (key.empty()) {
+	key = "ListMLE.model.default";
+    }
+    string model_data;
+    for (size_t i = 0; i < parameters.size(); ++i)
+	model_data += serialise_double(parameters[i]);
+    letor_db.set_metadata(key, model_data);
+}
+
+void
+ListMLERanker::load_model_from_metadata(const string& model_key)
+{
+    LOGCALL_VOID(API, "ListMLERanker::load_model_from_metadata", model_key);
+    Database letor_db(get_database_path());
+    string key = model_key;
+    if (key.empty()) {
+	key = "ListMLE.model.default";
+    }
+    string model_data = letor_db.get_metadata(key);
+    // Throw exception if no model data associated with key
+    if (model_data.empty()) {
+	throw LetorInternalError("No model found. Check key.");
+    }
```

Comment:
I wasn't talking about the exception string.  This is about **comments**, in particular the one in the code I commented on:

```
// Throw exception if no model data associated with key
```

The code following the comment clearly throws an exception if `model_data.empty()`, and `model_data` was read using key `key` on the line before the comment.  So what does this comment tell the reader that isn't clear from reading the code?